CONTRIBUTING.md

```diff
@@ -109,42 +115,346 @@ Check out [building instructions](https://github.com/prestodb/presto/tree/master
 
 Presto welcomes contributions from everyone.
 
-Contributions to Presto should be made in the form of GitHub pull request submissions and reviews. 
+Contributions to Presto should be made in the form of GitHub pull request submissions and reviews.
 
-Each pull request submission will be reviewed by a contributor or committer in the project.  In order for a PR to be eligible to be merged, a committer for the appropriate code must approve the code. Once approved by a committer, the PR may be merged by anyone. 
+Each pull request submission will be reviewed by a contributor or committer in the project.  In order for a PR to be eligible to be merged, a committer for the appropriate code must approve the code. Once approved by a committer, the PR may be merged by anyone.
 
 Pull request reviews are encouraged for anyone in the community who would like to contribute to Presto, and are
 expected from contributors and committers in at least equal proportion to their code contributions.
 
-Large contributions should have an associated GitHub issue.
+Contributions should have an associated GitHub issue.
+
+## Minimum Expectations for Contributing to Presto
+It is expected that everyone wishing to commit code will:
+* Work through the “Getting Started” materials
+* Read and agree to the Code of Conduct
+* Join Presto Slack
+* File an issue
+* Self verify new code matches codestyle for Presto
+* Follow commit standards
+* Add or modify existing tests related to code changes being submitted
+* Run and ensure local tests pass before submitting a merge request
+
+## Designing Your Code
+* Consider your code through 3 axes
+    1. Code Quality/Maintainability, e.g
+        1. Code Style
+        2. Maintainability - is the feature added in a hacky way (if we did everything that way would the code base become hard to use),  or is it implemented in a clean way with appropriate interfaces, fits in with the rest of Presto's design etc. (so keeps our code base maintainable).
+    1. Code Safety, e.g.
+        1. Adequate testing
+        1. Feature flags to enable/disable features that might be risky
+        1. Thread safe
+        1. No data structures that can grow without bounds. memory usage is accounted for
+        1. Not introducing expensive calls in a performance sensitive area
+    1. User friendliness e.g.
+        1. Config options have names and descriptions that can be understood by someone configuring Presto
+        1. All new language features, new functions, and major features have documentation added
+        1. Release notes following release note guidelines are added for user visible changes
+* For large features, discuss your design with relevant code owners before you start implementing it.
```

Comment:
Perhaps mention that large features may need an RFC: https://github.com/prestodb/rfcs

Reply:
added elsewhere with link to RFC